packages/opentelemetry-types/src/trace/span_data.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SpanContext } from './span_context';
+import { Resource } from '../resources/Resource';
+import { SpanKind } from './span_kind';
+import { Attributes } from './attributes';
+import { Link } from './link';
+import { Status } from './status';
+
+/**
+ * SpanData an object that is used to report out-of-band completed spans.
+ * The object and its members have to be treated as immutable.
+ */
+export interface SpanData {
```

Comment:
What would you think about having this just be a structural object type, so something like 

```
export interface SpanData {
   name: string;
   kind: SpanKind;
   ...
}
```

The advantage of a structural type is that you can easily create it and it also doesn't take extra JS code for the methods that would be added to the JS bundles (especially relevant for browsers)

Reply:
I did it like that because SpanData MUST be immutable.
So modeling it this way in TypeScript made sense to me.
But I am not a TypeScript wizard and I can also imagine simply using `readonly` for the properties. What do you think?

Reply:
What's the motivation for making SpanData immutable? My understanding is that is important for multi-threaded languages like Java/Go (looking at [Concurrency and Thread-Safety of OpenTelemetry API](https://github.com/open-telemetry/opentelemetry-specification/blob/d34318277b1999edccf5ab1d63024ec172ae5e49/specification/concurrency.md#concurrency-and-thread-safety-of-opentelemetry-api)).

However, in Node and the browser, due to the evented model, it's not actually possible for two threads to modify the same object, so the concerns about immutability for the sake of thread safety don't apply. (Multiple threads can mutate a SharedArrayBuffer, but that's a low level type, wheras objects get cloned when passed between say the browser UI thread and a WebWorker).

Would there be another reason for making this immutable besides thread safety?

Having the types be `readonly` is a nice way to document and enforce immutability if all consumers of the API used TypeScript, but for plain JS users of the API, the `readonly` just gets compiled away. So I agree that having getter methods or even `get` property accessors is the way to go if we need immutability.

Reply:
I agree with you but - correct me if I'm wrong - there are still benefits to immutability.
In our case, it's just a convention anyways.

SpanData is used to report out-of-band spans. The spec emphasizes the immutability of the members and also the members' contents if they are compound types.

Example:
SpanData is passed into the tracer.
Now let's assume that this is not immediately passed to the exporter but buffered somewhere.
Now let's say that `links` or `resource` is changed somwehere else while the Span has not yet been exported.
In this case, the contents of the object have changed and what is exported is not what it was when it was passed to the tracer.

The same applies if someone starts to modify the object itself after it has been passed to the tracer - like setting new links or another resource.
This would all lead to unexpected results when the export takes place.

This is why I kept the immutability rule from the spec. This may be a fallacy - I'm open to discuss.

Reply:
Thanks for explaining more of how this would be used! Makes sense on immutability making it more clear that you can't modify it after passing to the tracer, and that it keeps consistency with other languages.

I still have some lingering feelings of wishing this could just be a structure, but I think I can put those to rest particularly if others like this design. I also just learned about [Object.freeze](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze), and wonder if that could help us in some way.

Reply:
I stumbled over `Object.freeze` as well while looking into how to achieve immutability in JavaScript. We'll have to try when we do the Implementation, I guess.

I'm completely agnostic when it comes to defining just a POJO and no methods and I am still trying to figure out what to use when in this project. Sometimes we define accessors and no members and sometimes just the members. Is there some rule of thumb?

Reply:
How do you plan to use this interface? As per the current plan, we have decided to store `span`'s internal metadata/context in `LocalContext` object https://github.com/open-telemetry/opentelemetry-js/pull/63 (which also contains vendor related arbitrary metadata). I feel like `LocalContext` is kind of same as `SpanData`. Let me know if I am missing something.

Reply:
`SpanData` is specified [here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/tracing-api.md#recordspandata).

I need `SpanData` in `Tracer` as specified [here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/tracing-api.md#recordspandata).

Here is the [discussion](https://github.com/open-telemetry/opentelemetry-specification/pull/78#issuecomment-500024412) and Sergeys reasons to keep it like that for now.

Reply:
> I feel like LocalContext is kind of same as SpanData.

`SpanData` and `LocalContext` are 2 different things. `SpanData` is used only for exporters, while `LocalContext` is to be used within the tracer.

I strongly disagree with this design. It's a prime example of over-engineering. In its current state, it also doesn't provide a way to access the `LocalContext`, so how can vendor-specific properties be retrieved? Having only a single `SpanContext` to hold everything would be much simpler and easier to work with, both for vendors and for users. Especially since the 3 classes will basically hold the same kind of information and return them in different ways, which is unnecessary and confusing.

As @draffensperger mentioned, the main goals of immutability are specific to threaded languages, and there are no reasons to block the user from modifying the object in JavaScript if they want to do it. Immutability must be a hint and not enforced or it could block certain use cases.

Reply:
`SpanData` is passed to the tracer and it's part of the spec as is immutability.
Further above I describe how mutation can lead to unexpected results and this has nothing to do with thread safety.
I think all these API design discussions should be done within the specification SIG.
I don't find it efficient to discuss that again for every language and platform.

I don't mind implementation details. The only strong opinion I have is that if there is a spec we should stick to it or request a change if we disagree.

Reply:
The problem with a restrictive spec is that it cannot work cross-language. The spec should only describe concepts and a loose reference API. Since the current iteration is only for Java, we should experiment with an API that works better for JavaScript and then we'll be able to change the spec accordingly.

Also, an API that is 100% theoretical never works. We need to prove that it works, similar to the current Java work.

Reply:
> Further above I describe how mutation can lead to unexpected results and this has nothing to do with thread safety.

It can also prevent vendors to do certain things that they need. I think the idea is that APM vendors will be able to use the project implementation. If we end up needing to reimplement everything because of limitations that were added on purpose, then the value of OpenTelemetry becomes null. We can document that this should not be changed at a certain point in the lifecycle (mostly for users), but it should not be completely blocked.